components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/ServiceURLBuilderFactory.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.core;
+
+/**
+ * Service URL Builder Factory representation.
+ */
+public class ServiceURLBuilderFactory {
+
+    /**
+     * This method is called to create a {@link ServiceURLBuilder} instance.
+     *
+     * @return {@link ServiceURLBuilder}.
+     */
+    public ServiceURLBuilder getServiceURLBuilderInstance() {
+
+        return ServiceURLBuilder.create();
```

Comment:
Well the case is if somebody extends 'ServiceURLBuilder' still its create cannot be overloaded and the interface will return an implementation of DefaultURLServiceBuilder.

What if we consume the factory in ServiceURLBuilder.create() ?

Reply:
This should return new DefaultServiceURLBuilder();

and in create() of 'ServiceURLBuilder' you should have,

if IdentityCoreServiceComponent.getServiceURLBuilderFactory() != null {

return IdentityCoreServiceComponent.getServiceURLBuilderFactory().getServiceURLBuilder();

else 

new DefaultServiceURLBuilder ()